### DISCUSSION TITLE
Zoom ref's maxZoom value is 1 even if we set the maxZoomPixelRatio value as 4

### DISCUSSION THREAD

**Discussion by dineshkumarrajaguru** [igordanchenko/yet-another-react-lightbox#254] [General]
Hi @igordanchenko ,

I'm attempting to display a custom Zoom control on top of the lightbox but encountering an issue with enabling the zoom controls based on the Zoom level. During debugging, I discovered that the zoomRef value being passed to the custom control has a maxZoom value of 1 instead of the configured maxZoomPixelRatio value.

I've created a small [app](https://codesandbox.io/p/devbox/yarl-custom-zoom-68lj4n?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu9v9lbi00072v6in5pcctq5%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu9v9lbh00022v6it5ha1wwt%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu9v9lbh00042v6inobdmgy6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu9v9lbi00062v6ihkrvxe3n%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu9v9lbh00022v6it5ha1wwt%2522%253A%257B%2522id%2522%253A%2522clu9v9lbh00022v6it5ha1wwt%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clu9v9lbi00062v6ihkrvxe3n%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu9v9lbh00052v6i4vgoox7m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu9v9lbi00062v6ihkrvxe3n%2522%252C%2522activeTabId%2522%253A%2522clu9v9lbh00052v6i4vgoox7m%2522%257D%252C%2522clu9v9lbh00042v6inobdmgy6%2522%253A%257B%2522id%2522%253A%2522clu9v9lbh00042v6inobdmgy6%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu9v9lbh00032v6i4ac77puv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clu9ve3xa00702v6id0rrmecl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clu9ve43w00dydbip4el9e9di%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clu9v9lbh00032v6i4ac77puv%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) to replicate this issue. Could you please review it and provide insights on how to resolve this problem?

**Comment by igordanchenko** [comment]
Can you post a direct link, please? The shortened link you provided above does not work.

**Reply by dineshkumarrajaguru** [reply]
@igordanchenko I have updated the [link](https://codesandbox.io/p/devbox/yarl-custom-zoom-68lj4n?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu9v9lbi00072v6in5pcctq5%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu9v9lbh00022v6it5ha1wwt%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu9v9lbh00042v6inobdmgy6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu9v9lbi00062v6ihkrvxe3n%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu9v9lbh00022v6it5ha1wwt%2522%253A%257B%2522id%2522%253A%2522clu9v9lbh00022v6it5ha1wwt%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clu9v9lbi00062v6ihkrvxe3n%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu9v9lbh00052v6i4vgoox7m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu9v9lbi00062v6ihkrvxe3n%2522%252C%2522activeTabId%2522%253A%2522clu9v9lbh00052v6i4vgoox7m%2522%257D%252C%2522clu9v9lbh00042v6inobdmgy6%2522%253A%257B%2522id%2522%253A%2522clu9v9lbh00042v6inobdmgy6%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu9v9lbh00032v6i4ac77puv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clu9ve3xa00702v6id0rrmecl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clu9ve43w00dydbip4el9e9di%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clu9v9lbh00032v6i4ac77puv%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) now, could you please check now?

**Reply by igordanchenko** [reply]
Are you able to provide `width` and `height` attributes for your image slides? That should resolve the issue.

Here is what's going on here:
- your image slides do not supply `width` and `height` attributes
- during the initial render, Zoom plugin doesn't know what the `maxZoom` value is going to be (so it defaults to 1)
- image is loaded asynchronously
- image dimensions are obtained once the images is fully loaded
- Zoom context re-rerenders with the actual `maxZoom` value
- your custom zoom control does not re-render because it's using a ref obtained through the outer scope

There is a clean way to implement custom zoom controls through the `render.buttonZoom` render prop:

```jsx
render={{ buttonZoom: (zoomRef) => <CustomZoom zoomRef={zoomRef} /> }}
```

This way, your `CustomZoom` component is going to re-render anytime the `maxZoom` value changes. The downside of this approach is that this control must reside in the toolbar, and I'm not sure you can push it to the bottom of the lightbox as you currently do.

**Reply by dineshkumarrajaguru** [reply]
Indeed, when I specify the width and height for the Slides, the custom zoom controls function as intended. I didn't utilize the buttonZoom option because it positions the CustomZoom control in the toolbar (top-right).

I'm considering whether exposing the useZoom hook would allow us to access the configured values in the CustomZoom without needing to pass the references. Could you please clarify if my understanding is incorrect?

**Reply by igordanchenko** [reply]
Your understanding is correct. Exporting the `useZoom` hook from the plugin would be a better option. However, I ran into an issue with TypeScript type definitions while trying to export it, and I do not have a solution for that at the present time.

In the meantime, I can suggest the following workaround using React portal.

https://stackblitz.com/edit/yet-another-react-lightbox-254?file=src%2FApp.tsx

**Comment by dineshkumarrajaguru** [comment]
Thank you so much for resolving the issue for me again, I got the Controls at the bottom with proper zoom functionality without setting default height and width.
